mkosi/backend.py

```diff
@@ -471,10 +471,49 @@ def flatten(lists: Iterable[Iterable[T]]) -> list[T]:
     return list(itertools.chain.from_iterable(lists))
 
 
-def current_user_uid_gid() -> tuple[int, int]:
-    uid = int(os.getenv("SUDO_UID") or os.getenv("PKEXEC_UID") or os.getuid())
-    gid = pwd.getpwuid(uid).pw_gid
-    return uid, gid
+@dataclasses.dataclass
+class InvokingUser:
+    _pw: Optional[pwd.struct_passwd] = None
```

Comment:
I don't think this should be cached
```suggestion
    @staticmethod
    def _uid_from_env() -> Optional[int]
        if uid := (os.getenv("SUDO_UID") or os.getenv("PKEXEC_UID")):
            return int(uid)
        return None
```
this would also preclude this form being a dataclass and the properties would need to be methods, since class properties or static properties don't exist without hackery.

The rest would then become 
```python
@classmethod
def uid(cls) -> int:
    return _uid_from_env() or os.getuid()

Reply:
It's not cached, each call to current_user() will return a new instance of `InvokingUser`. `InvokingUser._pw` is an instance attribute, not a class attribute (as it's a dataclass).

Reply:
Well, it is cached in the instance. Of course, if you call the (alternative) constructor again, you will get a different instance. 

What I am proposing is just using a class as a namespace for a few associated functions instead of going through the whole dataclass machinery every time. I'll do a followup.

Reply:
> Well, it is cached in the instance. Of course, if you call the (alternative) constructor again, you will get a different instance.

But that's kind of the point. Once we have a uid (validated as existing in passwd), the other attributes are available and coherent with that uid. If I have some logic based on one attribute (the uid most of the time), the other attributes are guaranteed to be for the same uid.

Furthermore, apart from using the `SUDO_`* environment variables (but see https://github.com/systemd/mkosi/pull/1472#discussion_r1173669571), getting those attributes from the uid always goes through `pwd.getpwuid` anyway (directly or indirectly).

And dropping the alternative constructor can easily be done by directly using a default factory for `_pw`. If you want, I can create a new PR for that.

> What I am proposing is just using a class as a namespace for a few associated functions instead of going through the whole dataclass machinery every time. I'll do a followup.

But there is no dataclass machinery. An advantage of dataclasses is that it just generates the boilerplate code for the missing dunder methods when the class is decorated.

Anyway, I'm interested in your followup.